clash-prelude/src/Clash/Class/IndexTuple/Deriving.hs

```diff
@@ -0,0 +1,62 @@
+{-|
+Copyright  :  (C) 2023,      QBayLogic B.V.
+License    :  BSD2 (see the file LICENSE)
+Maintainer :  QBayLogic B.V. <devops@qbaylogic.com>
+-}
+
+{-# LANGUAGE TemplateHaskell #-}
+
+module Clash.Class.IndexTuple.Deriving
+  (IndexTuple(..), deriveIndexTupleInstances) where
+
+import Control.Monad.Extra (concatMapM)
+import Language.Haskell.TH
+
+class IndexTuple a where
+  -- | This function can be used to bind variables to numbers indexing
+  -- successive elements in a list-like structure. If the type of the indices
+  -- can be derived from context, it is enough to provide a tuple with the
+  -- variables like:
+  --
+  -- @
+  -- (a, b) = indexTuple
+  -- (a, b, c, d, e, f) = indexTuple
+  -- @
+  --
+  -- and so on. @a@ will be 0, @b@ 1, @c@ 2, etcetera. All variables will have
+  -- the same type, and specifying the type of any one of the variables is
+  -- enough. The type needs to be an instance of `Num`.
+  --
+  -- If a type needs to be specified here, you can use
+  --
+  -- @
+  -- (a :: Int, b, c) = indexTuple
+  -- @
+  --
+  -- or
+  --
+  -- @
+  -- a :: Int
+  -- (a, b, c) = indexTuple
```

Comment:
Also I'm a sucker for doctests :eyes:, any reason not to use that here?

Reply:
Syntax. I didn't like
```
>>> :{
... multiline
:}
```
much. But I can change it if you want.